controllers/helmrelease_chart_controller.go

```diff
@@ -0,0 +1,273 @@
+	}
+
+	// Confirm we are allowed to fetch the HelmChart.
+	if err := r.aclAllowAccessTo(obj, chartRef); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Attempt to fetch the HelmChart, ignoring not found errors.
+	var curChart sourcev1.HelmChart
+	err := r.Client.Get(ctx, chartRef, &curChart)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	// Build new HelmChart based on the declared template.
+	newChart := buildHelmChartFromTemplate(obj)
+
+	// HelmChart was not found, create it.
+	if apierrors.IsNotFound(err) {
+		if err = r.Client.Create(ctx, newChart); err != nil {
+			return ctrl.Result{}, err
+		}
+		obj.Status.HelmChart = chartRef.String()
+		r.Eventf(obj, events.EventSeverityTrace, "HelmChartCreated", "created HelmChart '%s' with SourceRef '%s/%s/%s'",
+			obj.Status.HelmChart, obj.Kind, obj.Namespace, obj.Name)
+		return ctrl.Result{RequeueAfter: obj.GetRequeueAfter()}, nil
+	}
+
+	// Check if the current spec diverges from the newly calculated spec.
+	if diff, eq := helmChartSpecDiff(curChart.Spec, newChart.Spec); !eq {
+		r.Eventf(obj, events.EventSeverityTrace, "HelmChartDrifted", "drift detected in HelmChart '%s':\n\n%s",
+			obj.Status.HelmChart, diff)
+
+		// Update the spec of the HelmChart.
+		curChart.Spec = newChart.Spec
+		if err = r.Client.Update(ctx, &curChart); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// From this moment on, we know the HelmChart spec is up-to-date.
+	obj.Status.HelmChart = chartRef.String()
+
+	// Requeue to ensure the state continues to be the same.
+	return ctrl.Result{RequeueAfter: obj.GetRequeueAfter()}, nil
+}
+
+// reconcileDelete handles the garbage collection of the current HelmChart in
+// the Status object of the given HelmRelease.
+func (r *HelmReleaseChartReconciler) reconcileDelete(ctx context.Context, obj *v2beta1.HelmRelease) (ctrl.Result, error) {
+	if obj.Status.HelmChart != "" {
+		ns, name := obj.Status.GetHelmChart()
+		namespacedName := types.NamespacedName{Namespace: ns, Name: name}
+
+		// Confirm we are allowed to fetch the HelmChart.
+		if err := r.aclAllowAccessTo(obj, namespacedName); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// Fetch the HelmChart.
+		var chart sourcev1.HelmChart
+		err := r.Client.Get(ctx, namespacedName, &chart)
+		if err != nil && !apierrors.IsNotFound(err) {
+			// Return error to retry until we succeed.
+			err = fmt.Errorf("failed to delete HelmChart '%s': %w", obj.Status.HelmChart, err)
+			return ctrl.Result{}, err
+		}
+		if err == nil {
+			// Delete the HelmChart.
+			if err = r.Client.Delete(ctx, &chart); err != nil {
+				err = fmt.Errorf("failed to delete HelmChart '%s': %w", obj.Status.HelmChart, err)
+				return ctrl.Result{}, err
+			}
+			r.Eventf(obj, events.EventSeverityTrace, "HelmChartDeleted", "deleted HelmChart '%s'", obj.Status.HelmChart)
+		}
+		// Truncate the chart reference in the status object.
+		obj.Status.HelmChart = ""
+	}
+
+	if obj.DeletionTimestamp != nil {
```

Comment:
This seems like a new check in all the various `reconcileDelete()` implementations.
Is this necessary in certain situations or just added it here for safety?

Reply:
It is new because `reconcileDelete` can happen both due to a change, and due to `Reconcile` calling it due to `DeletionTimestamp != nil`. Given this, it gate keeps a change from removing the finalizer.